src/metabase/util.clj

```diff
@@ -949,3 +949,10 @@
   [email-address domain]
   {:pre [(email? email-address)]}
   (= (email->domain email-address) domain))
+
+(defn field-ref->key
+  "A standard and repeatable way to address a column. Names can collide and sometimes are not unique. Field refs should
+  be stable, except we have to exclude the last part as extra information can be tucked in there. Names can be
+  non-unique at times, numeric ids are not guaranteed."
+  [field-ref]
+  (into [] (take 2) field-ref))
```

Comment:
I'm a little worried about this because `[:field 1 {:source-field 2}]` and `[:field 1 {:source-field 3}]` are two pretty different things. The former might mean `people.id via message.sender_id` while the latter means `people.id via message.receiver_id`.

Similarly, it's totally possible that you might include the same Field more than once in results with different temporal bucketing e.g. `[:field 1 {:temporal-unit :year}]` and `[:field 1 {:temporal-unit :month}]`. In some cases that distinction is important but in other cases it's not -- I've gotten tripped up on this a lot in the past...

Exactly _what_ is relevant and what isn't for key purposes isn't something I've entirely figured out TBH.

Reply:
Another example: it's possible that some column got included in the final results multiple times as the result of different joins, in which case the `:join-alias` is relevant... editing the metadata for one shouldn't affect the other

Reply:
FWIW it's not like this would have worked in the past where we were just using IDs instead of Field refs, so I wouldn't necessarily block this PR on fixing this issue, but it is something to be aware of that is at least worth a mention in comments in the code

Reply:
yeah i agonized about this. If the only thing that matters is positional then this is pretty difficult. I was hoping that the field-ref way would work but was nervous. If positional is the only way to match these things up, it seems like we cannot handle changes in the query and also persist changes in the metadata.

Reply:
I actually was trying to solve a similar problem and the approach I eventually went with was to use _some_ but not _all_ of the options in the `:field` clause for key purposes. See `metabase.query-processor.middleware.add-implicit-clauses/source-metadata->fields`

https://github.com/metabase/metabase/blob/master/src/metabase/query_processor/middleware/add_implicit_clauses.clj#L53-L64

I don't think that would 100% fix the problem tho because we're trying to merge together two things. Maybe instead we need some smart code that will try to find the closest matching Field ref. e.g. if the User-supplied metadata has a `:field_ref` of `[:field 1 nil]` we'll first look for result metadata with Field ref `[:field 1 nil]`, and failing that we'll look for `[:field 1 _]` or something. Idk

Reply:
I don't think positional is a good idea since Tables can get new columns or have columns dropped

Reply:
agree on positional being basically dead in the water.

Reply:
Actually, one thing that _should_ (almost always) be unambiguous is `:name`. The only case where a query can come back with multiple columns with the same `:name` is in native queries, but those are explicitly disallowed from being used as the source of another query (and hence wouldn't be allowed for datasets anyway)

Reply:
I considered something like that. But it would also need to exclude base-types. Those are only present for native after the query runs and we update the metadata since we cannot compute them at the start like we can in mbql.

Reply:
Hmmm... `:base-type` really isn't important anyway for key purposes. It's mostly so we can apply the appropriate QP functions when using a native query as a source query. I think `(select-keys options [:source-field :join-alias])` would _probably_ be sufficient for almost everything. Unless you included a Field twice with multiple bucketing options. In that case I guess it's just 💥 😠 🔫